### PR TITLE
[WIP] Animation API Docs

### PR DESCRIPTION
**Pull request by mincrmatt12**
This is _NOT_ finished yet, but I thought I would create the PR now so I can get some feedback on this.

### CODE REVIEW

File: docs/animation/intro.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+read the documentation on blockstates
+
+
+First, some terms:
+
+| Term | Description |
+| --- | --------- |
+| Model File | Your JSON file to be animated |
+| Armature File | The JSON file which defines the joints and clips of a model |
+| Clip | A collection of variables to apply to certain joints. Sort of an "animation group" |
```
Comment (williewillus) [inline docs/animation/intro.md]: not accurate. clip = set of transformations that can be applied to a joint

File: docs/animation/intro.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+| Joint | A name for a json cube |
```
Comment (williewillus) [inline docs/animation/intro.md]: not accurate: joint = any part of the model you want to consider as one cohesive unit for animation purposes. doesn't have to be the json system, as for example b3d models can have joints defined too

File: docs/animation/intro.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+| ASM File | Defines how to use the armature file to animate the model |
+| Parameter | Some value that is used to control what part of a clip we are on |
+| State | A state for the animation to be in, controls what clips are active |
+| Transitions | Controls which states go into what other states |
+| Event | An event caused by animation, readable by the thing being animated |
```
Comment (williewillus) [inline docs/animation/intro.md]: How about "A way to receive notifications in-code when the animation reaches a certain point"

File: docs/animation/intro.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+First, some terms:
+
+| Term | Description |
+| --- | --------- |
+| Model File | Your JSON file to be animated |
+| Armature File | The JSON file which defines the joints and clips of a model |
+| Clip | A collection of variables to apply to certain joints. Sort of an "animation group" |
+| Joint | A name for a json cube |
+| ASM File | Defines how to use the armature file to animate the model |
```
Comment (williewillus) [inline docs/animation/intro.md]: this seems really vague

File: docs/animation/armature.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+ "joints": {
+  "joint_name": {"index in model file": [1.0]}
+ }
+```
+
+- `joint_name` is the name of the joint
+- `index in model file` is a 0-indexed number denoting which model element this joint controls
+
+Not all elements need to have a joint, only the ones you are animating.
+
```
Comment (williewillus) [inline docs/animation/armature.md]: maybe add a note that the `[1.0]` is required(?) but unused

File: docs/animation/armature.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+- offset_x, offset_y, offset_z - translation
+- scale - uniform scaling
+- scale_x, scale_y, scale_x - scaling on certain axes
+- axis_x, axis_y, axis_z - rotation axes
+- angle - rotation angle
+
+and if #3875 is merged, there will be:
+
+- origin_x, origin_y, origin_z - rotation origin
+
```
Comment (williewillus) [inline docs/animation/armature.md]: missing explanation for `samples`, and where the input values for joint clips come from

File: docs/animation/asm.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+### Clips
+
+!!! note
+    
+    Clips can either be ASM-clips, ones that are defined in the ASM, or armature-clips, ones that are defined in the armature file.
+    For the rest of this page, "clips" will refer to ASM-clips unless otherwise stated
+
+A clip controls what a state will do. They can be one of many types, the easiest to use being the `TimeClip`. These animate a given clip by a given parameter.
+Other kinds include the `IdentityClip`, which does nothing, the `ClipReference`, which should "only be used for debugging" and just does the same as another clip,
```
Comment (williewillus) [inline docs/animation/asm.md]: `"only be used for debugging"` what?

**Comment by williewillus** [comment]
For all of your json examples, I would recommend making something like an expanded grammar,
 like [this](https://gist.github.com/williewillus/9ebb0d04329526e31564) but more explanations.

also recommend using javascript on the syntax highlighter instead of json so it won't complain about as many things (like // comments which aren't valid json but are valid javascript)

**Comment by williewillus** [comment]
Good start at documenting the monster, though :P

File: docs/animation/armature.md (hunk 5 above)
Comment (mincrmatt12) [inline docs/animation/armature.md]: It has something to do with scale, got to check the source

File: docs/animation/asm.md (hunk 7 above)
Comment (mincrmatt12) [inline docs/animation/asm.md]: it says that in the source code, idk

File: docs/animation/armature.md (hunk 5 above)
Comment (williewillus) [inline docs/animation/armature.md]: fry said it was unused, something about blending the joint clips/TRSR's together

File: docs/animation/asm.md (hunk 7 above)
Comment (williewillus) [inline docs/animation/asm.md]: when writing documentation, "idk" is not a good thing to be saying ;p

I think this is the code-side deserialization of a `#` reference in the ASM json, that's what the comment means. So it's not that you should only use it for debugging, it's just that you'll find it rare to construct one in code instead of specifying it in json

File: docs/animation/armature.md [hunk 8]
```diff
@@ -0,0 +1,147 @@
+Armature Files
+==============
+
+Armature Files define joints and clips for animating a model.
+
+File Structure
+----------------
+
+An example armature file, taken from the forge debug mod (requires #3875)
```
Comment (elifoster) [inline docs/animation/armature.md]: What is this number referring to?

File: docs/animation/armature.md [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+### Events
+
+Each clip can fire events, formatted like this:
+```json
+"events": {
+  "event_time": "event_text"
+}
+```
+For more information about events and what `event_text` means, see the page on ASMs.
```
Comment (elifoster) [inline docs/animation/armature.md]: Link to the actual ASM page here.

File: docs/animation/asm.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
+ASM Files
```
Comment (elifoster) [inline docs/animation/asm.md]: Change this to "Animation State Machine Files".

File: docs/animation/asm.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+===========
+
+ASM Files are the meat of the animation API. These define how the animation is carried out and how to use the clips defined in the armature file.
```
Comment (elifoster) [inline docs/animation/asm.md]: Change this to "Animation State Machine (ASM) Files are the meat..."

File: docs/animation/intro.md [hunk 12]
```diff
@@ -0,0 +1,30 @@
+Intro to the Animation API
+===============================
+
+The Forge Animation API lets you animate json (and B3D) models.
```
Comment (elifoster) [inline docs/animation/intro.md]: `json` -> `JSON`

File: docs/animation/intro.md (hunk 4 above)
Comment (elifoster) [inline docs/animation/intro.md]: Link to Animation State Machine page here perhaps

File: docs/animation/armature.md (hunk 8 above)
Comment (mincrmatt12) [inline docs/animation/armature.md]: forge pr

File: docs/animation/intro.md (hunk 2 above)
Comment (mincrmatt12) [inline docs/animation/intro.md]: i plan to cover b3d-specific stuff at the end of the docs, because very few mods i've seen actually use b3d models.

File: docs/animation/armature.md (hunk 5 above)
Comment (mincrmatt12) [inline docs/animation/armature.md]: looking in the code it takes the final transformation and multiplies it by that

Comment (mincrmatt12) [inline docs/animation/armature.md]: looking at the b3d specification, it appears to be put there because b3d has it

**Comment by mincrmatt12** [comment]
@williewillus IMO for a beginner who just wants to animate a model a grammar file seems a bit over the top, but i'll include them anyways

File: docs/animation/intro.md (hunk 2 above)
Comment (pWn3d1337) [inline docs/animation/intro.md]: I think .obj is important as there are conversion tools for techne/tabula models and those were common in 1.7.

Comment (mincrmatt12) [inline docs/animation/intro.md]: AFAIK the API doesn't actually support .obj models

Comment (pWn3d1337) [inline docs/animation/intro.md]: Are you sure on that? That's a would be a problem, because forge supports loading of json, obj and b3d. So it should work on all 3. If not, we need to open another issue :-/.

Comment (mincrmatt12) [inline docs/animation/intro.md]: It doesn't, because obj has no builtin way of specifying the joints. You could do it by implementing a lot of the animation interfaces for OBJ stuff, but then you would need forge specific extensions. Or some way to define meshes with the semi-standard "g" tag

Comment (pWn3d1337) [inline docs/animation/intro.md]: That's really bad. I have to look for a b3d converter.

https://en.wikipedia.org/wiki/Wavefront_.obj_file the o and g tags could work.

**Comment by wiiv** [comment]
Hello, I've been playing around with this and I'd like to know if the "origin" variables are already implemented.

**Comment by tterrag1098** [comment]
What's left to be done here? I would really love for this to be merged soon.

File: docs/animation/armature.md [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+  }
+}
+
+```
+
+The file is organized in two sections, joints and clips.
+
+Joints
+--------
+Each joint defines a connection to the model file. A joint can animate any number of elements in the model file, but they are all animated with the same input value and transforms
```
Comment (williewillus) [inline docs/animation/armature.md]: "a connection to the model file" feels a bit ambiguous. How about

`A joint defines an object that animates together as a single unit in the animation together. For example, in the vanilla json model format, multiple elements can belong to the same joint.`

File: docs/animation/armature.md [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+
+<joint_definition> ::= {
+    <string>: [ <float> ] // index_model, joint_weight
+}
+
+```
+
+- `joint_name` is the name of the joint
+- `index_model` is a 0-indexed number denoting a model element this joint controls. Must be a string (see example)
+- `joint_weight` is a weight (0-1) of how much this element will animate. The transformation matrix caused by a clip animating this element will be multiplied by this number.
```
Comment (williewillus) [inline docs/animation/armature.md]: according to fry a while ago the weights are unused, is that still the case?

also by "transformation matrix will be multiplied by this number" what does that mean, the transform is lerped, or the whole matrix is just scalarly multiplied, or what?

File: docs/animation/armature.md [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+    "variable": <variable>,
+    "type": "uniform",
+    "interpolation": <interpolation>,
+    "samples": [ float, ... ]
+}
+
+
+```
+
+- loop: whether or not the parameter value can continue increasing and the animation loops, or if it stops at the ending state
```
Comment (williewillus) [inline docs/animation/armature.md]: not sure what "parameter value can continue increasing and the animation loops" means, the grammar isn't quite clear

File: docs/animation/armature.md [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+
+- offset_x, offset_y, offset_z - translation
+- scale - uniform scaling
+- scale_x, scale_y, scale_x - scaling on certain axes
+- axis_x, axis_y, axis_z - rotation axes
+- angle - rotation angle
+
+and if PR #3875 is merged, there will be:
+
+- origin_x, origin_y, origin_z - rotation origin
```
Comment (williewillus) [inline docs/animation/armature.md]: the pr got merged (i think)

File: docs/animation/armature.md [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+
+### Events
+
+Each clip can fire events, formatted like this:
+```javascript
+<event> :: {
+    <event_time>: "event_text"
+}
+```
+For more information about events and what `event_text` means, see the page on ASMs.
```
Comment (williewillus) [inline docs/animation/armature.md]: use the full name here and link said page

File: docs/animation/armature.md [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+
+
+```
+
+- loop: whether or not the parameter value can continue increasing and the animation loops, or if it stops at the ending state
+
+### Joint Clips
+Each `joint_clip` is a set of variables to change for a joint. The `type` attribute is currently ignored, but must be `"uniform"`.
+
+`samples` defines what value the animation will take on, and its interpretation depends on the value of `interpolation`.
```
Comment (williewillus) [inline docs/animation/armature.md]: i don't think this really captures what samples really means, but unfortunately I can't think of something better.
Maybe `samples defines the boundary and granularity of values used for the transform`? meh

File: docs/animation/asm.md [hunk 19]
```diff
@@ -0,0 +1,312 @@
...
+
+An example:
+```java
+@Nullable
+private final IAnimationStateMachine asm;
+private final VariableValue cycle = new VariableValue(4);
+
+public Spin() {
+    if (FMLCommonHandler.instance().getSide() == Side.CLIENT) {
+        asm = ModelLoaderRegistry.loadASM(new ResourceLocation(MODID, "asms/block/rotatest.json"), ImmutableMap.of("cycle_length", cycle));
```
Comment (williewillus) [inline docs/animation/asm.md]: is this actually side safe? i'm not sure how loading of methods works but we don't want to show dediserver-crashing code in the official docs

File: docs/animation/asm.md [hunk 20]
```diff
@@ -0,0 +1,312 @@
...
+- `r`: output = floor(input / arg) * arg;
+- `R`: output = ceil(input / arg) * arg;
+- `f`: output = input - floor(input / arg) * arg;
+- `F`: output = ceil(input / arg) * arg - input;
+
+##### Example explanations:
+- input + 4
+- (input / 5) + 1
+- input + 2 + value of parameter other
+- input + value of parameter other + value of parameter cycle given input 3
```
Comment (williewillus) [inline docs/animation/asm.md]: I'm reading this md file unrendered but maybe move these to be right under their respective examples?

File: docs/animation/asm.md [hunk 21]
```diff
@@ -0,0 +1,312 @@
...
+
+- blend the closed clip to the open clip, giving both clips the unaltered time as input and blend progress `#progress`.
+- blend the result of the move clip when given the input parameter `mover` to the end clip with the unaltered time as the input with blend progress `#progress`.
+
+### States
+
+The states section of the file is simply a list of all possible states.
+For example 
+```json
+"states": {
```
Comment (williewillus) [inline docs/animation/asm.md]: this should be a json array right?

**Comment by mincrmatt12** [comment]
sorry for the delay, (life happens), i'm going to be working on this more in the coming days, and hopefully finishing it

File: docs/animation/asm.md (hunk 19 above)
Comment (mincrmatt12) [reply docs/animation/asm.md]: actually you are right, this should be proxy'd (oops)

File: docs/animation/armature.md (hunk 16 above)
Comment (mincrmatt12) [reply docs/animation/armature.md]: it did indeed

File: docs/animation/armature.md (hunk 18 above)
Comment (mincrmatt12) [reply docs/animation/armature.md]: its more like keyframes in traditional animation

File: docs/animation/armature.md (hunk 14 above)
Comment (mincrmatt12) [reply docs/animation/armature.md]: i remember seeing it being used it ModelBlockAnimation, lines 545-570.

File: docs/animation/asm.md (hunk 20 above)
Comment (mincrmatt12) [reply docs/animation/asm.md]: i think it might be better this was, so you can see some examples, get the explanation of the entire parameter type, and then you can check to see the explanations of the examples

File: docs/animation/armature.md (hunk 13 above)
Comment (mincrmatt12) [reply docs/animation/armature.md]: sound good to me

**Comment by ichttt** [comment]
I don't think the abbreviate Animation State Machine to ASM should be used (or at least the page should be  renamed), as the term is already used by forge with a different meaning, and most people assoisiate assemly bytecode with the term, and forge has things called ASMDataTable, ASMEventHandler, ASMTransformer and so on. But I would like some other people's feedback on that

**Comment by mincrmatt12** [comment]
You have a point, but saying Animation State Machine every time is very verbose. Perhaps putting a warning /note admonition at the top would work

**Comment by Anime-Quatroctus** [comment]
implementing.md "Your block must also render in the ENTITYBLOCK_ANIMATED render layer." This is not true, if the BlockState has Properties.StaticProperty. And the TESR is an AnimationTESR it will get the model from the static=false variant and use that instead. Everything else looks amazing, what else is there left to do?

**Comment by mincrmatt12** [comment]
All that needs to be done is finish the documentation on implementing the api for items and entities. Also thanks for telling me about the block stuff, most of what I know is from reading the source and experimentation.

File: docs/animation/implementing.md [hunk 22]
```diff
@@ -0,0 +1,42 @@
+Implementing the API
+======================
+
+Depending on what you want to animate with the API, code-side implementation is a bit different.
+Documentation on the ASM api itself (for controlling the animation) is found on the [ASM][asm] page because it is independent of what
+you are animation.
```
Comment (MajorTuvok) [inline docs/animation/implementing.md]: Shouldn't it be animating?

File: docs/animation/implementing.md [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+{
+    @Override
+    public void handleEvents(Chest chest, float time, Iterable<Event> pastEvents)
+    {
+        chest.handleEvents(time, pastEvents);
+    }
+}); 
+```
+
+In this example, we've overrided the `handleEvents()` callback when we registered the TESR because the implementation is simple, but you could easily subclass
```
Comment (MajorTuvok) [inline docs/animation/implementing.md]: And maybe 'overridden' here?

**Comment by MajorTuvok** [comment]
(I'm no native speaker (and especially unsure about the correct 3.form of (to) override, but I'm pretty sure it is not the standard - ed?). Please correct me if I'm wrong...)

File: docs/animation/implementing.md (hunk 22 above)
Comment (Anime-Quatroctus) [reply docs/animation/implementing.md]: Please make a thread on http://www.minecraftforge.net/forum/forum/70-modder-support/

Comment (MajorTuvok) [reply docs/animation/implementing.md]: ???
I just wanted to help with spelling mistakes... 
I wasn't even asking questions...

Comment (Anime-Quatroctus) [reply docs/animation/implementing.md]: Im sorry I misinterpreted what you said and posted. Yes that should be "animating" instead of animation.

Comment (MajorTuvok) [reply docs/animation/implementing.md]: No Problem...
I modified it in hopes of making it clearer to others, to ehat I'm referring to.

File: docs/animation/implementing.md (hunk 23 above)
Comment (mincrmatt12) [reply docs/animation/implementing.md]: Yes, you are correct. Fixed.

File: docs/animation/implementing.md (hunk 22 above)
Comment (mincrmatt12) [reply docs/animation/implementing.md]: Fixed.

File: docs/animation/implementing.md [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+you are animating.
+
+Blocks
+--------
+
+Animations for blocks are done with the AnimationTESR, which is a FastTESR. Because of this, having a TileEntity for your block
+is necessary. Your TileEntity must provide the `ANIMATION_CAPABILITY`, which is recieved by calling its `.cast` method with your
+ASM. Your block must also render in the `ENTITYBLOCK_ANIMATED` render layer if you do not provide a `StaticProperty` in the block's blockstate.
+
+The `StaticProperty` is a property you can add to your block's blockstate by adding `Properties.StaticProperty` to the list of your block's properties inside
```
Comment (PaleoCrafter) [inline docs/animation/implementing.md]: Although it's kind of obvious from the name, you should mention that all parts of the model that *can* be static should probably also be rendered in the static state, that's the whole point.

File: docs/animation/implementing.md [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+    public void handleEvents(Chest chest, float time, Iterable<Event> pastEvents)
+    {
+        chest.handleEvents(time, pastEvents);
+    }
+}); 
+```
+
+In this example, we've overridden the `handleEvents()` callback when we registered the TESR because the implementation is simple, but you could easily subclass
+AnimationTESR to achieve the same effect. `handleEvents()` callback for blocks takes two arguments: the tile entity being rendered, and an iterable of the events.
+The call to ``chest.handleEvents`` calls a method located in the fictional `Chest` TileEntity, as the ASM is not accessible inside of the handleEvents method. is not accessible inside of the handleEvents method.
```
Comment (PaleoCrafter) [inline docs/animation/implementing.md]: Accidentally repeated yourself, it seems. `handleEvents` in backticks please, and I think you've got one to many on each side of `chest.handleEvents`.

File: docs/animation/asm.md [hunk 26]
```diff
@@ -0,0 +1,307 @@
...
+The ASM contains _parameters_, _clips_, _states_, and _transitions_. The simplest ones to explain are states and transitions, so I'll start with those
+
+### States
+
+The Animation _State_ Machine can be in many different _states_. You define which states there are in the states section.
+Each _clip_ implements a state, telling the ASM how to animate it.
+
+### Transitions
+
+Right now, the only uses for transitions are to define which states can go to which other states, no way to make a transition-only
```
[the remark this answers is not in the document]
Comment (mincrmatt12) [reply docs/animation/asm.md]: I realize that, but from the perspective of someone who just wants to animate something this definition makes it clear that you are not able to say, define an animation that plays _during_ a transition between two states

File: docs/animation/asm.md [hunk 27]
```diff
@@ -0,0 +1,307 @@
...
+
+Different types of parameters have different formats for `<parameter_definition>`, and the simple ones are:
+
+- `IdentityValue`: the string `#identity`,
+- `ParameterValue`: the parameter to reference, prefixed with `#`, e.g. `#my_awesome_parameter`
+- `ConstValue`: a number to use as the constant to return
+
+Other kinds are a bit more complicated
+
+#### `SimpleExprValue`
```
[the remark this answers is not in the document]
Comment (mincrmatt12) [reply docs/animation/asm.md]: In my opinion, it makes more sense to leave the names as they are in the code since being able to look up the implementation may be useful for people.

File: docs/animation/asm.md (hunk 20 above)
Comment (mincrmatt12) [reply docs/animation/asm.md]: The idea I had with the examples was to first show examples, then explain the parameter, and then show the examples again with their meaning now that the reader has read the explanation.

File: docs/animation/asm.md [hunk 28]
```diff
@@ -0,0 +1,307 @@
...
+```json
+[ "compose", "#cycle", 3]
+[ "compose", "#test", "#other"]
+[ "compose", [ "+", 3], "#other"]
+[ "compose", [ "compose", "#other2", "#other3"], "#other"]
+```
+##### Explanation
+
+`CompositionValue` takes two parameter definitions as inputs, and does `value1(value2(input))`. In other words, it chains
+the two inputs, calling the second one with the given input, and the first one with the output of the second one.
```
[the remark this answers is not in the document]
Comment (mincrmatt12) [reply docs/animation/asm.md]: If the reader knows about that, they wouldn't need the more detailed explanation.

File: docs/animation/asm.md [hunk 29]
```diff
@@ -0,0 +1,307 @@
...
+}
+```
+
+As with parameters, different kinds of clips have different formats for `<clip_definition>`, but the simple ones are:
+
+- `IdentityClip`: the string `#identity`
+- `ClipReference`: the clip name prefixed with `#`, e.g. `#my_amazing_clip`
+- `ModelClip`: a model resource location + `@` + the name of the armature-clip, e.g. `mymod:block/test@default` or `mymod:block/test#facing=east@moving`
+
+#### `TimeClip`
```
[the remark this answers is not in the document]
Comment (mincrmatt12) [reply docs/animation/asm.md]: And again I would rather keep the names for easy searching, as it applies more here as creating custom clips may be required for adding animations to custom models.

File: docs/animation/implementing.md (hunk 25 above)
Comment (mincrmatt12) [reply docs/animation/implementing.md]: Curse the incompatibilities between reStructuredText and Markdown!

File: docs/animation/intro.md [hunk 30]
```diff
@@ -0,0 +1,28 @@
...
+!!! note
+    
+    Although you can use B3D models with the Animation API, most of this documentation will assume you
+    are using JSON files. (TODO) See the page on using B3D models for more information
+
+The Animation API is made up of two main components: armature files and animation state machine (ASM) files.
+Armature files define joints and clips for JSON files. Joints are names for cubes in the model file with weights, while
+clips are a set of transformations to apply to joints. ASM files define the various states an animation can be in, as well as what
+transitions exist between those states. They also define the parameters for an animation, generally used as inputs to clips, but can also
+trigger events. Events are essentially a way to recieve in-code notifications when the animation reaches a certain point, but can also trigger
```
[the remark this answers is not in the document]
Comment (mincrmatt12) [reply docs/animation/intro.md]: There's an option in the armature file to make a clip automatically loop, so value outside of the 0-1 range have their integer part removed to create a looping animation, so the input to a clip is not always 0-1

**Comment by mincrmatt12** [comment]
Thanks for the input, I've tried to address the changes that you're requesting.

File: docs/animation/asm.md (hunk 26 above)
[the remark this answers is not in the document]
Comment (PaleoCrafter) [reply docs/animation/asm.md]: Ah, I understand where you're coming from now. Maybe say that a transition is not an animation unto itself or a continuous change between two states but instead an instantaneous switch to a new state. Also mention that these smooth transitions have to be modelled by an extra state, as you point out yourself that that might be a source of confusion the elimination of which is the entire point of documenting this.

File: docs/animation/asm.md (hunk 27 above)
[the remark this answers is not in the document]
Comment (PaleoCrafter) [reply docs/animation/asm.md]: It's okay to mention the implementation name (in parentheses maybe), but the primary focus should be on the high level concepts and this for example represents mathematical expressions for useful calculations, while `SimpleExprValue` doesn't carry much information.

Besides, most people reading these docs won't ever see any of the parameter implementation classes except for `VariableValue`. The point of documenting this is to not have to look at the implementation.

File: docs/animation/asm.md (hunk 28 above)
[the remark this answers is not in the document]
Comment (PaleoCrafter) [reply docs/animation/asm.md]: I'm not saying drop the explanation, just to mention the comparison for people who *are* familiar with the concept. It's also helpful if people maybe don't understand your explanation and want to look it up themselves.

File: docs/animation/asm.md (hunk 27 above)
[the remark this answers is not in the document]
Comment (mincrmatt12) [reply docs/animation/asm.md]: Ok, I see what you mean.

**Comment by mincrmatt12** [comment]
Alright, now I think I fixed the issues with class names

**Comment by mincrmatt12** [comment]
Are there any more changes that need to be made before this is merged?
